src/audio/dai-zephyr.c

```diff
@@ -956,17 +820,64 @@ int dai_common_params(struct dai_data *dd, struct comp_dev *dev,
 		 * frame_fmt's (using pcm converter).
 		 */
 		hw_params.frame_fmt = dev->ipc_config.frame_fmt;
-		buffer_set_params(dd->dma_buffer, &hw_params,
-				  BUFFER_UPDATE_FORCE);
-
+		buffer_set_params(dd->dma_buffer, &hw_params, BUFFER_UPDATE_FORCE);
 		dd->sampling = get_sample_bytes(hw_params.frame_fmt);
 	}
 
 	dd->fast_mode = dd->ipc_config.feature_mask & BIT(IPC4_COPIER_FAST_MODE);
+	return 0;
+}
+
+int dai_common_params(struct dai_data *dd, struct comp_dev *dev,
+		      struct sof_ipc_stream_params *params)
+{
+	struct dma_sg_config *config = &dd->config;
+	uint32_t period_bytes = 0;
+	uint32_t period_count = 0;
+	int err = 0;
+
+	comp_dbg(dev, "dai_zephyr_params()");
 
-	return dev->direction == SOF_IPC_STREAM_PLAYBACK ?
-		dai_playback_params(dd, dev, period_bytes, period_count) :
-		dai_capture_params(dd, dev, period_bytes, period_count);
+	/* configure dai_data first */
+	err = ipc_dai_data_config(dd, dev);
+	if (err < 0) {
+		comp_err(dev, "dai_zephyr_params(): ipc dai data config failed.");
+		return err;
+	}
+
+	err = dai_verify_params(dd, dev, params);
+	if (err < 0) {
+		comp_err(dev, "dai_zephyr_params(): pcm params verification failed.");
+		return -EINVAL;
+	}
+
+	err = dai_set_dma_buffer(dd, dev, params, &period_bytes, &period_count);
+	if (err < 0) {
+		comp_err(dev, "dai_zephyr_params(): alloc dma buffer failed.");
+		goto out;
+	}
+
+	err = dai_set_sg_config(dd, dev, period_bytes, period_count);
+	if (err < 0) {
+		comp_err(dev, "dai_zephyr_params(): set sg config failed.");
+		goto out;
+	}
+
+	err = dai_set_dma_config(dd, dev);
+	if (err < 0)
+		comp_err(dev, "dai_zephyr_params(): set dma config failed.");
+out:
+	/*
+	 * Make sure to free all allocated items, all functions
+	 * can be called with null pointer.
+	 */
+	if (err < 0) {
+		buffer_free(dd->dma_buffer);
+		dma_sg_free(&config->elem_array);
+		rfree(dd->z_config);
```

Comment:
if I understand it correctly, `dd` is persistent, isn't it? Here in error case we free several buffers, but we keep pointers to them. Should we reset them all to `NULL`? Except `config->elem_array.elems` - it's already reset inside `dma_sg_free()`

Reply:
dd is allocated by dai_new and released in dai_free. If dai_common_params is failed,  dai_reset will be called and these pointer will be set to null
````c
void dai_common_reset(struct dai_data *dd, struct comp_dev *dev)
{
	struct dma_sg_config *config = &dd->config;

	dma_sg_free(&config->elem_array);
	if (dd->z_config) {
		rfree(dd->z_config->head_block);
		rfree(dd->z_config);
		dd->z_config = NULL;
	}

	if (dd->dma_buffer) {
		buffer_free(dd->dma_buffer);
		dd->dma_buffer = NULL;
	}
}
````

Reply:
@RanderWang that's even worse then. That would mean that we free the same pointers twice, which is wrong and can lead to run-time errors

Reply:
o, thanks!